text/0000-collection-recovery.md

```diff
@@ -0,0 +1,173 @@
+    // removed nickel
+    // > add 2 cobalt
+    // added cobalt
+    // > add 3 zinc
+    // replaced copper with zinc
+    // ```
+    //
+    // However, `HashSet` does not expose its items via its `{contains, insert, remove}` methods,
+    // instead providing only a boolean indicator of the item's presence in the set, preventing us
+    // from implementing the desired functionality.
+}
+```
+
+# Detailed design
+
+Add the following item-recovery methods to `std::collections::{BTreeSet, HashSet}`:
+
+```rust
+impl<T> Set<T> {
+    // Like `contains`, but returns a reference to the item if the set contains it.
+    fn item<Q: ?Sized>(&self, item: &Q) -> Option<&T>;
+
+    // Like `remove`, but returns the item if the set contained it.
+    fn remove_item<Q: ?Sized>(&mut self, item: &Q) -> Option<T>;
+
+    // Like `insert`, but replaces the item with the given one and returns the previous item if the
+    // set contained it.
+    fn replace(&mut self, item: T) -> Option<T>;
+}
+```
+
+In order to implement the above methods, add the following key-recovery methods to
+`std::collections::{BTreeMap, HashMap}`:
+
+```rust
+impl<K, V> Map<K, V> {
+    // Like `get`, but additionally returns a reference to the entry's key.
+    fn key_value<Q: ?Sized>(&self, key: &Q) -> Option<(&K, &V)>;
+
+    // Like `get_mut`, but additionally returns a reference to the entry's key.
+    fn key_value_mut<Q: ?Sized>(&mut self, key: &Q) -> Option<(&K, &mut V)>;
+
+    // Like `remove`, but additionally returns the entry's key.
+    fn remove_key_value<Q: ?Sized>(&mut self, key: &Q) -> Option<(K, V)>;
+
+    // Like `insert`, but additionally replaces the key with the given one and returns the previous
+    // key and value if the map contained it.
+    fn replace(&mut self, key: K, value: V) -> Option<(K, V)>;
+}
+```
+
+For completion, add the following key-recovery methods to
+`std::collections::{btree_map, hash_map}::OccupiedEntry`:
+
+```rust
+impl<'a, K, V> OccupiedEntry<'a, K, V> {
+    // Like `get`, but additionally returns a reference to the entry's key.
+    fn key_value(&self) -> (&K, &V);
+
+    // Like `get_mut`, but additionally returns a reference to the entry's key.
+    fn key_value_mut(&mut self) -> (&K, &mut V);
+
+    // Like `into_mut`, but additionally returns a reference to the entry's key.
+    fn into_key_value_mut(self) -> (&'a K, &'a mut V);
+
+    // Like `remove`, but additionally returns the entry's key.
+    fn remove_key_value(self) -> (K, V);
+}
+```
+
+# Drawbacks
+
+This complicates the collection APIs.
+
+The distinction between `insert` and `replace` may be confusing. It would be more consistent to
+call `Set::replace` `Set::insert_item` and `Map::replace` `Map::insert_key_value`, but `BTreeMap`
+and `HashMap` do not replace equivalent keys in their `insert` methods, so rather than have
+`insert` and `insert_key_value` behave differently in that respect, `replace` is used instead.
+
+# Alternatives
+
+Do nothing.
```

Comment:
It's not possible to use `HashMap` that way, because it doesn't provide any methods that return `&K` (or `K`) other than via its iterators.